### PR TITLE
Transaction meta data cleanup

### LINKED ISSUE #26: Define transaction description and user to be unicode
See:

https://groups.google.com/forum/#!topic/python-transaction/Yn326XwCZ5E

### PR DESCRIPTION
**Pull request by jimfulton**
This fixes #26 and #27.

- The transaction ``user`` and ``description`` attributes are now  defined to be text (unicode) as opposed
  to Python the ``str`` type.

- Added the ``extended_info`` transaction attribute which contains
  transaction meta data.  (The ``_extension`` attribute is retained as
  an alias for backward compatibility.)

  The transaction interface, ``ITransaction``, now requires
  ``extended_info`` keys to be text (unicode) and values to be
  JSON-serializable.

### CODE REVIEW

**Comment by jimfulton** [comment]
Note that I manually tested this with ZODB 4 and 5 under Python 2.7 and 3.5.

File: CHANGES.rst [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+- The transaction ``user`` and ``description`` attributes are now
+  defined to be text (unicode) as apposed to Python the ``str`` type.
+
+- Added the ``extended_info`` transaction attribute which contains
+  transaction meta data.  (The ``_extension`` attribute is retained as
+  an alias for backward compatibility.)
+
+  The transaction interface, ``ITransaction``, now requires
+  ``extended_info`` keys to be text (unicode) and values to be
+  JSON-serializable.
```
Comment (tseaver) [inline CHANGES.rst]: Because these are new features and potential backward-incompatibilities, ISTM we should be bumping the target next version to `2.0` with this PR, and maybe calling out the backward-incompatibility more explicitly.

File: transaction/_transaction.py [hunk 2]
```diff
@@ -118,6 +118,10 @@ def __init__(self, synchronizers=None, manager=None):
         # List of (hook, args, kws) tuples added by addAfterCommitHook().
         self._after_commit = []
 
+    @property
+    def _extension(self):
+        return self.extended_info
```
Comment (tseaver) [inline transaction/_transaction.py]: Add a comment noting the BBB purpose?

File: transaction/_transaction.py [hunk 3]
```diff
@@ -504,19 +508,19 @@ def note(self, text):
...
         """ See ITransaction.
         """
-        self.user = "%s %s" % (path, user_name)
+        self.user = u"%s %s" % (path, user_name)
 
     def setExtendedInfo(self, name, value):
         """ See ITransaction.
         """
-        self._extension[name] = value
+        self.extended_info[name + u''] = value
```
Comment (tseaver) [inline transaction/_transaction.py]: I guess that might be faster than a manual typecheck, but the purpose is a bit unclear.

File: transaction/_transaction.py [hunk 4]
```diff
@@ -504,19 +508,19 @@ def note(self, text):
...
-            self.description += "\n" + text
+            self.description += u"\n" + text
         else:
             self.description = text
 
     def setUser(self, user_name, path="/"):
         """ See ITransaction.
         """
-        self.user = "%s %s" % (path, user_name)
+        self.user = u"%s %s" % (path, user_name)
```
Comment (tseaver) [inline transaction/_transaction.py]: If `setUser` is no longer part of the `ITransaction` interface, maybe we should be making `user` (and `description` into properties which do typechecking?

File: transaction/interfaces.py [hunk 5]
```diff
@@ -176,21 +183,17 @@ def note(text):
...
 
-    def setUser(user_name, path="/"):
-        """Set the user name.
-
-        path should be provided if needed to further qualify the
-        identified user.  This is a convenience method used by Zope.
-        It sets the .user attribute to str(path) + " " + str(user_name).
-        This sets the `.user` attribute; see its docs for more detail.
-        """
-
```
Comment (tseaver) [inline transaction/interfaces.py]: The changelog doesn't call out dropping `setUser` from the interface:  is that change intended?

File: CHANGES.rst (hunk 1 above)
Comment (jimfulton) [inline CHANGES.rst]: OK I think I'll revert that change.  I don't really need it.

Comment (jimfulton) [inline CHANGES.rst]: OK, then I'll leave it and make a backward incompatibility statement.

File: transaction/_transaction.py (hunk 2 above)
Comment (jimfulton) [inline transaction/_transaction.py]: OK

File: transaction/_transaction.py (hunk 4 above)
Comment (jimfulton) [inline transaction/_transaction.py]: OK

File: transaction/_transaction.py (hunk 3 above)
Comment (jimfulton) [inline transaction/_transaction.py]: It's also less tedious. I'll add a comment to clarify

File: transaction/interfaces.py (hunk 5 above)
Comment (jimfulton) [inline transaction/interfaces.py]: I'll clarify. My plan is to keep it forever, but to drop it from ITransaction because it's unseemly :)

**Comment by jimfulton** [comment]
I think I've addressed all of the comments.

Thanks!

**Comment by jimfulton** [comment]
@tseaver you cool?

**Comment by jimfulton** [comment]
This turned out to be more disruptive than I expected, for a non-obvious reason

ZEO uses transaction objects to hold transaction data.  This data may already be bytes rather than unicode.  For example, the ZEO Python 3.5 recovery tests now fail because when copying data, the data are bytes and on Python 3, there isn't an automatic conversion. The tests would fail on Python 2 too if the tests used non-ascii meta data.  I'm going to fix this in ZEO.  Hopefully the problem isn't more widespread.  It's encouraging that the tests are passing for the built-in storages.

This reminds me a bit of WSGI, where it's not always clear at what level encoding should happen.

**Comment by jimfulton** [comment]
Oh look, worms, in cans! :)

``ZODB.interfaces.IStorageTransactionInformation`` is pythonically vague about types and I'd be willing to bet $.05 that all storages implement these as bytes.  But these are used in ``ZODB.BaseStorage.copy`` and passed to ``ZODB.interfaces.IStorage.tpc_begin``.

``ZODB.interfaces.IStorage.tpc_begin`` doesn't say what it expects its ``transaction`` argument to be. (Of course, I'm self mocking here.)  The storage implementations are defensive WRT meta data types, which is probably find for now.  Maybe we'll want to clean this up some day.

(I realize I'm mostly talking to myself. Writing things down helps me think. :))

**Comment by jamadden** [comment]
For the record, this also has some minor impact in RelStorage.

For RelStorage 2.0, it's mostly trivial; the only real breakage is in tests under Python 2. The database storage is in bytes, but due to the way Python 3 was handled, those fields were already allowed to be either text or bytes on either Python 2 or 3.  One corner case is that the transaction iterator was always returning that data as bytes, though, in the TransactionRecord, so I guess that needs to change to decoding them back to text under transaction 2. (I see @jimfulton has added a comment about IStorageTransactionInformation here, which is TransactionRecord.)

RelStorage 1.x will have problems with this change if the user/description contains non-default-decodable characters (it calls `str(trans.user)`, etc). But I guess you shouldn't use transaction 2 with RelStorage 1.

**Comment by jimfulton** [comment]
IMO, storages, or maybe ZODB should commit to an encoding.

*Maybe* encoding should be done by ZODB itself.

If we want to, we could specify the data passed to tpc_begin as bytes.

IMO, it would be much cleaner if storages always dealt with bytes.

**Comment by jamadden** [comment]
Another place these are used is in `IStorage.undoLog` (that's a breakage I missed in my first pass over the RelStorage test case).

**Comment by jimfulton** [comment]
I have a slightly similar issue with byteserver and extension data.

Life would be simpler for me if extension data were always pickled before passing them to storages.

**Comment by jimfulton** [comment]
Moving this conversation to the zodb list :)
